drivers/src/dma.rs

```diff
@@ -0,0 +1,253 @@
+            }
+        });
+
+        Ok(())
+    }
+
+    fn dma_write_fifo(&mut self, write_data: &[u8]) -> CaliptraResult<()> {
+        let dma = self.dma.regs_mut();
+
+        let max_fifo_depth = dma.cap().read().fifo_max_depth();
+        let current_fifo_depth = dma.status0().read().fifo_depth();
+
+        if write_data.len() as u32 > max_fifo_depth - current_fifo_depth {
+            return Err(CaliptraError::DRIVER_DMA_FIFO_OVERRUN);
+        }
+
+        write_data.chunks(4).for_each(|word| {
+            let ptr = dma.write_data().ptr as *mut u8;
+            // Reg only exports u32 writes but we need finer grained access
+            unsafe {
+                ptr.copy_from_nonoverlapping(word.as_ptr(), word.len());
+            }
+        });
+
+        Ok(())
+    }
+
+    fn do_transaction(&mut self) -> CaliptraResult<()> {
+        let dma = self.dma.regs_mut();
+
+        let status0 = dma.status0().read();
+        if status0.busy() {
+            return Err(CaliptraError::DRIVER_DMA_TRANSACTION_ALREADY_BUSY);
+        }
+
+        if status0.error() {
+            return Err(CaliptraError::DRIVER_DMA_TRANSACTION_ERROR);
+        }
+
+        dma.ctrl().modify(|c| c.go(true));
+
+        while dma.status0().read().busy() {
+            if dma.status0().read().error() {
+                return Err(CaliptraError::DRIVER_DMA_TRANSACTION_ERROR);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn read_dword(&mut self, read_addr: usize) -> CaliptraResult<u32> {
+        let mut read_val: u32 = 0;
+
+        self.flush();
+
+        let read_transaction = DmaReadTransaction {
+            read_addr,
+            fixed_addr: false,
+            length: core::mem::size_of::<u32>() as u32,
+            target: DmaReadTarget::AhbFifo,
+        };
+
+        self.setup_dma_read(read_transaction);
+        self.do_transaction()?;
+        self.dma_read_fifo(read_val.as_bytes_mut())?;
+        Ok(read_val)
+    }
+
+    pub fn write_dword(&mut self, write_addr: usize, write_val: u32) -> CaliptraResult<()> {
+        self.flush();
+
+        let write_transaction = DmaWriteTransaction {
+            write_addr,
+            fixed_addr: false,
+            length: core::mem::size_of::<u32>() as u32,
+            origin: DmaWriteOrigin::AhbFifo,
+        };
+        self.dma_write_fifo(write_val.as_bytes())?;
+        self.setup_dma_write(write_transaction);
+        self.do_transaction()?;
+        Ok(())
+    }
+
+    pub fn transfer_payload_to_mbox(
+        &mut self,
+        read_addr: usize,
+        payload_len_bytes: u32,
+        fixed_addr: bool,
+        block_size: u32,
+    ) -> CaliptraResult<()> {
+        self.flush();
+
+        let read_transaction = DmaReadTransaction {
+            read_addr,
+            fixed_addr,
+            length: payload_len_bytes,
+            target: DmaReadTarget::Mbox,
+        };
+        self.setup_dma_read(read_transaction);
+        self.dma
```

Comment:
We should add a description comment to the function stating that the client needs to take the mailbox lock before calling this function.